File: scripts/bibtex-to-yaml.py
#!/usr/bin/env python3
"""
Converts a BibTeX entry to a portal resource YAML file.

Usage:
    python scripts/bibtex-to-yaml.py input.bib --projects APRICOT PHASES
    cat input.bib | python scripts/bibtex-to-yaml.py --projects APRICOT
    python scripts/bibtex-to-yaml.py input.bib --projects ODFA --pubmed https://pubmed.ncbi.nlm.nih.gov/12345/ --output data/resources/pub-example.yaml

Options:
    --projects      Project names (e.g. APRICOT PHASES)
    --pubmed        PubMed URL
    --link          Additional link as "Label - URL" (can be repeated)
    --description   Override the generated description
    --output        Output file path (default: prints to stdout)

Requirements:
    pip install bibtexparser pyyaml
"""

import argparse
import re
import sys
from datetime import date

try:
    import bibtexparser
except ImportError:
    print("Error: bibtexparser is required. Install it with: pip install bibtexparser", file=sys.stderr)
    sys.exit(1)

try:
    import yaml
except ImportError:
    print("Error: pyyaml is required. Install it with: pip install pyyaml", file=sys.stderr)
    sys.exit(1)


MONTH_MAP = {
    "jan": "01", "january": "01",
    "feb": "02", "february": "02",
    "mar": "03", "march": "03",
    "apr": "04", "april": "04",
    "may": "05",
    "jun": "06", "june": "06",
    "jul": "07", "july": "07",
    "aug": "08", "august": "08",
    "sep": "09", "september": "09",
    "oct": "10", "october": "10",
    "nov": "11", "november": "11",
    "dec": "12", "december": "12",
}


def parse_month(month_str: str) -> str | None:
    if not month_str:
        return None
    m = month_str.strip().lower()
    if m.isdigit():
        return m.zfill(2)
    return MONTH_MAP.get(m)


def clean_latex(text: str) -> str:
    """Remove common LaTeX artifacts from a string."""
    text = re.sub(r"[{}]", "", text)
    text = text.replace("\\&", "&")
    text = text.replace("\\textendash", "\u2013")
    text = text.replace("\\textemdash", "\u2014")
    text = text.replace("~", " ")
    text = re.sub(r"\\[a-zA-Z]+\s*", "", text)  # strip remaining commands
    text = re.sub(r"\s+", " ", text).strip()
    return text


def make_id(key: str, title: str) -> str:
    base = key if key else title
    slug = re.sub(r"[^a-z0-9]+", "-", base.lower()).strip("-")[:60]
    return f"pub-{slug}"


def extract_entry(bibtex_str: str) -> dict:
    parser = bibtexparser.bparser.BibTexParser(common_strings=True)
    library = bibtexparser.loads(bibtex_str, parser=parser)
    if not library.entries:
        print("Error: No BibTeX entries found in input.", file=sys.stderr)
        sys.exit(1)
    return library.entries[0]


def build_resource(entry: dict, args) -> dict:
    title = clean_latex(entry.get("title", "Untitled"))
    entry_key = entry.get("ID", "")
    resource_id = make_id(entry_key, title)

    # Date
    year = entry.get("year")
    month_str = entry.get("month")
    pub_year = int(year) if year else None
    pub_month = None
    if month_str:
        m = parse_month(month_str)
        if m:
            pub_month = int(m)

    # DOI
    doi = None
    if "doi" in entry:
        doi_val = entry["doi"].strip()
        doi = doi_val if doi_val.startswith("http") else f"https://doi.org/{doi_val}"

    # Venue
    venue = None
    for key in ("journal", "booktitle", "publisher", "howpublished"):
        if key in entry:
            venue = clean_latex(entry[key])
            break

    # Description
    if args.description:
        description = args.description
    elif "abstract" in entry:
        abstract = clean_latex(entry["abstract"])
        first_sentence = re.match(r"^[^.!?]+[.!?]", abstract)
        description = first_sentence.group(0) if first_sentence else abstract[:200] + "..."
    elif venue:
        description = f"{title[:100]}. Published in {venue}."
    else:
        description = title

    # Links
    links = []
    if doi:
        links.append({"label": "DOI", "url": doi, "platform": "Website"})
    if args.pubmed:
        links.append({"label": "PubMed", "url": args.pubmed, "platform": "Website"})
    for extra in (args.link or []):
        parts = extra.split(" - ", 1)
        if len(parts) == 2:
            links.append({"label": parts[0].strip(), "url": parts[1].strip(), "platform": "Website"})

    # Tags
    tags = ["paper"]
    if venue:
        venue_tag = re.sub(r"[^a-z0-9]+", "-", venue.lower()).strip("-")[:30]
        tags.append(venue_tag)

    # Projects
    projects = args.projects or []

    resource = {
        "id": resource_id,
        "name": title,
        "type": "Publication",
    }
    if pub_year:
        resource["publishedYear"] = pub_year
    if pub_month:
        resource["publishedMonth"] = pub_month
    resource["description"] = description
    resource["developedByProjects"] = projects if projects else []
    resource["usedByProjects"] = []
    resource["links"] = links if links else []
    resource["tags"] = tags
    resource["status"] = "Active"
    resource["lastUpdated"] = date.today().isoformat()

    return resource


def resource_to_yaml(resource: dict) -> str:
    """Produce clean YAML output with controlled field order."""
    field_order = [
        "id", "name", "type", "publishedYear", "publishedMonth",
        "description", "developedByProjects", "usedByProjects", "links",
        "tags", "status", "lastUpdated",
    ]

    lines = []
    for key in field_order:
        if key not in resource:
            continue
        value = resource[key]

        if isinstance(value, list) and len(value) == 0:
            lines.append(f"{key}: []")
        elif isinstance(value, list) and key == "links":
            lines.append(f"{key}:")
            for link in value:
                lines.append(f"  - label: {link['label']}")
                lines.append(f"    url: {link['url']}")
                lines.append(f"    platform: {link['platform']}")
        elif isinstance(value, list):
            lines.append(f"{key}:")
            for item in value:
                lines.append(f"  - {item}")
        elif isinstance(value, str) and any(c in value for c in ":\"'#[]{}&*!|>%@`"):
            lines.append(f'{key}: "{value}"')
        else:
            lines.append(f"{key}: {value}")

    return "\n".join(lines) + "\n"


def main():
    parser = argparse.ArgumentParser(
        description="Convert a BibTeX entry to a portal resource YAML file."
    )
    parser.add_argument("input", nargs="?", help="BibTeX input file (reads stdin if omitted)")
    parser.add_argument("--projects", nargs="+", help="Project names (e.g. APRICOT PHASES)")
    parser.add_argument("--pubmed", help="PubMed URL")
    parser.add_argument("--link", action="append", help='Additional link as "Label - URL"')
    parser.add_argument("--description", help="Override the generated description")
    parser.add_argument("--output", help="Output file path")

    args = parser.parse_args()

    if args.input:
        with open(args.input) as f:
            bibtex_str = f.read()
    elif not sys.stdin.isatty():
        bibtex_str = sys.stdin.read()
    else:
        parser.print_help()
        sys.exit(1)

    entry = extract_entry(bibtex_str)
    resource = build_resource(entry, args)
    yaml_str = resource_to_yaml(resource)

    if args.output:
        with open(args.output, "w") as f:
            f.write(yaml_str)
        print(f"Written to {args.output}", file=sys.stderr)
    else:
        sys.stdout.write(yaml_str)
        suggested = f"data/resources/{resource['id']}.yaml"
        print(f"\nSuggested file: {suggested}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: scripts/bibtex-to-yaml.js
#!/usr/bin/env node

/**
 * Converts a BibTeX entry to a portal resource YAML file.
 *
 * Usage:
 *   node scripts/bibtex-to-yaml.js input.bib [--projects APRICOT,PHASES] [--pubmed URL] [--output path]
 *   cat input.bib | node scripts/bibtex-to-yaml.js --projects APRICOT
 *
 * Options:
 *   --projects   Comma-separated project names (e.g. APRICOT,BSO-AD)
 *   --pubmed     PubMed URL
 *   --link       Additional link (format: "Label - URL")
 *   --output     Output file path (default: prints to stdout)
 *   --description  Override the generated description
 */

const fs = require("fs");
const path = require("path");

// --- BibTeX parser (handles common cases) ---

function parseBibtex(input) {
  const entry = {};

  // Extract entry type and key
  const typeMatch = input.match(/@(\w+)\s*\{([^,]*),/);
  if (typeMatch) {
    entry._type = typeMatch[1].toLowerCase();
    entry._key = typeMatch[2].trim();
  }

  // Extract fields — handles braced {values} and quoted "values"
  // Also handles multi-line values
  const fieldRegex = /(\w+)\s*=\s*(?:\{((?:[^{}]|\{[^{}]*\})*)\}|"([^"]*)"|(\d+))/g;
  let match;
  while ((match = fieldRegex.exec(input)) !== null) {
    const key = match[1].toLowerCase();
    const value = (match[2] || match[3] || match[4] || "").trim();
    // Clean up LaTeX artifacts
    entry[key] = value
      .replace(/\\\&/g, "&")
      .replace(/\\textendash/g, "–")
      .replace(/\\textemdash/g, "—")
      .replace(/[{}]/g, "")
      .replace(/\s+/g, " ")
      .trim();
  }

  return entry;
}

// --- Extract structured data ---

function extractData(entry) {
  const title = entry.title || "Untitled";

  // Build ID from key or title
  const id =
    "pub-" +
    (entry._key || title)
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .slice(0, 60);

  // Date
  let publishedDate = null;
  if (entry.year) {
    const month = parseMonth(entry.month);
    publishedDate = month ? `${entry.year}-${month}` : `${entry.year}-01`;
  }

  // DOI
  let doi = null;
  if (entry.doi) {
    doi = entry.doi.startsWith("http")
      ? entry.doi
      : `https://doi.org/${entry.doi}`;
  }

  // Description
  let description = null;
  if (entry.abstract) {
    // Use first sentence or first 200 chars of abstract
    const firstSentence = entry.abstract.match(/^[^.!?]+[.!?]/);
    description = firstSentence
      ? firstSentence[0].trim()
      : entry.abstract.slice(0, 200).trim() + "...";
  }

  // Journal / venue
  const venue =
    entry.journal ||
    entry.booktitle ||
    entry.publisher ||
    entry.howpublished ||
    null;

  // Authors
  const authors = entry.author || null;

  return { id, title, publishedDate, doi, description, venue, authors };
}

function parseMonth(monthStr) {
  if (!monthStr) return null;
  const m = monthStr.toLowerCase().trim();

  // Numeric
  if (/^\d+$/.test(m)) return m.padStart(2, "0");

  // Month names/abbreviations
  const months = {
    jan: "01", january: "01",
    feb: "02", february: "02",
    mar: "03", march: "03",
    apr: "04", april: "04",
    may: "05",
    jun: "06", june: "06",
    jul: "07", july: "07",
    aug: "08", august: "08",
    sep: "09", september: "09",
    oct: "10", october: "10",
    nov: "11", november: "11",
    dec: "12", december: "12",
  };
  return months[m] || null;
}

// --- Generate YAML ---

function generateYaml(data, options) {
  const description =
    options.description ||
    data.description ||
    (data.venue
      ? `${data.title.slice(0, 100)}. Published in ${data.venue}.`
      : data.title);

  const projects = options.projects || [];

  const lines = [];
  lines.push(`id: ${data.id}`);
  lines.push(`name: ${yamlString(data.title)}`);
  lines.push(`type: Publication`);
  if (data.publishedDate) {
    lines.push(`publishedDate: "${data.publishedDate}"`);
  }
  lines.push(`description: ${yamlString(description)}`);

  // Projects
  if (projects.length > 0) {
    lines.push(`developedByProjects:`);
    projects.forEach((p) => lines.push(`  - ${p}`));
  } else {
    lines.push(`developedByProjects: []`);
  }
  lines.push(`usedByProjects: []`);

  // Links
  const links = [];
  if (data.doi) {
    links.push({ label: "DOI", url: data.doi, platform: "Website" });
  }
  if (options.pubmed) {
    links.push({ label: "PubMed", url: options.pubmed, platform: "Website" });
  }
  if (options.extraLinks) {
    options.extraLinks.forEach((l) => {
      const [label, url] = l.split(" - ").map((s) => s.trim());
      if (label && url) {
        links.push({ label, url, platform: "Website" });
      }
    });
  }

  if (links.length > 0) {
    lines.push(`links:`);
    links.forEach((link) => {
      lines.push(`  - label: ${link.label}`);
      lines.push(`    url: ${link.url}`);
      lines.push(`    platform: ${link.platform}`);
    });
  } else {
    lines.push(`links: []`);
  }

  // Tags
  const tags = ["paper"];
  if (data.venue) {
    tags.push(
      data.venue
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-|-$/g, "")
        .slice(0, 30)
    );
  }
  lines.push(`tags:`);
  tags.forEach((t) => lines.push(`  - ${t}`));

  lines.push(`status: Active`);

  const today = new Date().toISOString().slice(0, 10);
  lines.push(`lastUpdated: "${today}"`);

  return lines.join("\n") + "\n";
}

function yamlString(s) {
  if (/[:\#\[\]{}&*!|>'"%@`]/.test(s) || s.includes("\n")) {
    return `"${s.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`;
  }
  return s;
}

// --- CLI ---

function main() {
  const args = process.argv.slice(2);
  let inputFile = null;
  const options = { projects: [], extraLinks: [] };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--projects":
        options.projects = args[++i].split(",").map((s) => s.trim());
        break;
      case "--pubmed":
        options.pubmed = args[++i];
        break;
      case "--link":
        options.extraLinks.push(args[++i]);
        break;
      case "--output":
        options.output = args[++i];
        break;
      case "--description":
        options.description = args[++i];
        break;
      default:
        if (!args[i].startsWith("-")) inputFile = args[i];
    }
  }

  // Read input
  let input;
  if (inputFile) {
    input = fs.readFileSync(inputFile, "utf-8");
  } else if (!process.stdin.isTTY) {
    input = fs.readFileSync(0, "utf-8");
  } else {
    console.error(
      "Usage: node scripts/bibtex-to-yaml.js input.bib [--projects APRICOT,PHASES] [--pubmed URL]"
    );
    console.error("       cat input.bib | node scripts/bibtex-to-yaml.js --projects APRICOT");
    process.exit(1);
  }

  const entry = parseBibtex(input);
  if (!entry.title) {
    console.error("Error: Could not parse a title from the BibTeX input.");
    process.exit(1);
  }

  const data = extractData(entry);
  const yaml = generateYaml(data, options);

  if (options.output) {
    fs.writeFileSync(options.output, yaml);
    console.error(`Written to ${options.output}`);
  } else {
    process.stdout.write(yaml);
    // Suggest filename
    const suggestedFile = `data/resources/${data.id}.yaml`;
    console.error(`\nSuggested file: ${suggestedFile}`);
  }
}

main();
